homeassistant/components/iskra/__init__.py

```diff
@@ -50,50 +63,49 @@ async def async_setup_entry(hass: HomeAssistant, entry: IskraConfigEntry) -> boo
         )
     elif conf[CONF_PROTOCOL] == "rest_api":
         authentication = None
-        if conf.get(CONF_USERNAME) or conf.get(CONF_PASSWORD):
+        if (username := conf.get(CONF_USERNAME)) is not None and (
+            password := conf.get(CONF_PASSWORD)
+        ):
             authentication = {
-                "username": conf[CONF_USERNAME],
-                "password": conf[CONF_PASSWORD],
+                "username": username,
+                "password": password,
             }
         adapter = RestAPI(ip_address=conf[CONF_HOST], authentication=authentication)
 
     # Try connecting to the device and create pyiskra device object
     try:
-        root_device = await Device.create_device(adapter)
+        base_device = await Device.create_device(adapter)
     except DeviceConnectionError as e:
-        _LOGGER.error("Cannot connect to the device: %s", e)
-        return False
+        raise ConfigEntryNotReady("Cannot connect to the device") from e
     except NotAuthorised as e:
-        _LOGGER.error("Not authorised: %s", e)
-        return False
+        raise ConfigEntryNotReady("Not authorised to connect to the device") from e
     except DeviceNotSupported as e:
-        _LOGGER.error("Device not supported: %s", e)
-        return False
+        raise ConfigEntryNotReady("Device not supported") from e
 
     # Initialize the device
-    await root_device.init()
+    await base_device.init()
 
     # if the device is a gateway, add all child devices, otherwise add the device itself.
-    if root_device.is_gateway:
+    if base_device.is_gateway:
         # Add the gateway device to the device registry
         device_registry = dr.async_get(hass)
         device_registry.async_get_or_create(
             config_entry_id=entry.entry_id,
-            identifiers={(DOMAIN, root_device.serial)},
+            identifiers={(DOMAIN, base_device.serial)},
             manufacturer=MANUFACTURER,
-            name=f"{root_device.serial}",
-            model=root_device.model,
-            sw_version=root_device.fw_version,
+            name=f"{base_device.serial}",
+            model=base_device.model,
+            sw_version=base_device.fw_version,
         )
 
         coordinators = [
             IskraDataUpdateCoordinator(hass, child_device)
-            for child_device in root_device.get_child_devices()
+            for child_device in base_device.get_child_devices()
         ]
     else:
-        coordinators = [IskraDataUpdateCoordinator(hass, root_device)]
+        coordinators = [IskraDataUpdateCoordinator(hass, base_device)]
 
-    entry.runtime_data = {"base_device": root_device, "coordinators": coordinators}
+    entry.runtime_data = IskraRuntimeData(coordinators)
```

Comment:
We should run `coordinator.async_config_entry_first_refresh()` to make sure we have data in the coordinator

Reply:
OK, moved it here from sensor.py makes more sense